src/fix/main.c

```diff
@@ -65,535 +75,944 @@ static void print_usage(void)
-		 * This is a bitmap image that displays when the Game Boy is
-		 * turned on. It must be intact, or the game will not boot.
-		 */
-
-		/*
-		 * See also: global checksums at 0x14D–0x14F, They must
-		 * also be correct for the game to boot, so we fix them
-		 * as well when requested with the -f flag.
-		 */
-
-		uint8_t ninlogo[48] = {
-			0xCE, 0xED, 0x66, 0x66, 0xCC, 0x0D, 0x00, 0x0B,
-			0x03, 0x73, 0x00, 0x83, 0x00, 0x0C, 0x00, 0x0D,
-			0x00, 0x08, 0x11, 0x1F, 0x88, 0x89, 0x00, 0x0E,
-			0xDC, 0xCC, 0x6E, 0xE6, 0xDD, 0xDD, 0xD9, 0x99,
-			0xBB, 0xBB, 0x67, 0x63, 0x6E, 0x0E, 0xEC, 0xCC,
-			0xDD, 0xDC, 0x99, 0x9F, 0xBB, 0xB9, 0x33, 0x3E
-		};
-
-		if (trashlogo) {
-			for (int i = 0; i < sizeof(ninlogo); i++)
-				ninlogo[i] = ~ninlogo[i];
-		}
+	va_start(ap, fmt);
+	vfprintf(stderr, fmt, ap);
+	va_end(ap);
 
-		memcpy(header + 0x04, ninlogo, sizeof(ninlogo));
-	}
+	if (nbErrors != UINT8_MAX)
+		nbErrors++;
+}
 
-	if (settitle) {
-		/*
-		 * Offset 0x134–0x143: Game Title
-		 * This is a sixteen-character game title in ASCII (no high-
-		 * bit characters).
-		 */
-
-		/*
-		 * See also: CGB flag at 0x143. The sixteenth character of
-		 * the title is co-opted for use as the CGB flag, so they
-		 * may conflict.
-		 */
-
-		/*
-		 * See also: Game ID at 0x13F–0x142. These four ASCII
-		 * characters may conflict with the title.
-		 */
-
-		strncpy((char *)header + 0x34, title, 16);
-	}
+static const uint8_t ninLogo[] = {
+	0xCE, 0xED, 0x66, 0x66, 0xCC, 0x0D, 0x00, 0x0B,
+	0x03, 0x73, 0x00, 0x83, 0x00, 0x0C, 0x00, 0x0D,
+	0x00, 0x08, 0x11, 0x1F, 0x88, 0x89, 0x00, 0x0E,
+	0xDC, 0xCC, 0x6E, 0xE6, 0xDD, 0xDD, 0xD9, 0x99,
+	0xBB, 0xBB, 0x67, 0x63, 0x6E, 0x0E, 0xEC, 0xCC,
+	0xDD, 0xDC, 0x99, 0x9F, 0xBB, 0xB9, 0x33, 0x3E
+};
 
-	if (setid) {
-		/*
-		 * Offset 0x13F–0x142: Game ID
-		 * This is a four-character game ID in ASCII (no high-bit
-		 * characters).
-		 */
+static enum { DMG, BOTH, CGB } model = DMG; // If DMG, byte is left alone
+#define   FIX_LOGO        0x80
+#define TRASH_LOGO        0x40
+#define   FIX_HEADER_SUM  0x20
+#define TRASH_HEADER_SUM  0x10
+#define   FIX_GLOBAL_SUM  0x08
+#define TRASH_GLOBAL_SUM  0x04
+static uint8_t fixSpec = 0;
+static const char *gameID = NULL;
+static uint8_t gameIDLen;
+static bool japanese = true;
+static const char *newLicensee = NULL;
+static uint8_t newLicenseeLen;
+static uint16_t oldLicensee = 0x100;
+static enum MbcType cartridgeType = MBC_NONE;
+static uint16_t romVersion = 0x100;
+static uint16_t padValue = 0x100;
+static uint16_t ramSize = 0x100;
+static bool sgb = false; // If false, SGB flags are left alone
+static const char *title = NULL;
+static uint8_t titleLen;
+
+static uint8_t maxTitleLen(void)
+{
+	return gameID ? 11 :  model != DMG ? 15 : 16;
+}
 
-		memcpy(header + 0x3F, id, 4);
+static ssize_t readBytes(int fd, void *buf, size_t len)
+{
+	// POSIX specifies that lengths greater than SSIZE_MAX yield implementation-defined results
+	assert(len <= SSIZE_MAX);
```

Comment:
I would turn this into an "if (x) exit(1)" or something like that. If someone builds a release build (which is what we expect people to do) and they hit this corner case (which is _extremely_ unlikely), we will have no information of what went wrong.

Reply:
I'm expecting such cases to be forwarded to us, and using a debug build *would* fix the issue. It's an internal error anyway, it cannot provide an end user any valuable info.

Reply:
So you're leaving the assert because maybe some user hits it and reports it, but users can't hit the assert because they will use release builds.

Reply:
I mean that they would hit the behavior that the assert checks, report weird behavior to us, and we'd see the assert.

Reply:
Well, but the user could be finding the issue in a project that is closed source and doesn't want to share with you, and maybe it's not easy to get to a small example that fails the assert. In any case, I doubt this specific assert will ever be a problem... I'm just not a fan of assert(), but I do like static_assert() a lot.

Reply:
To put it differently: this `assert` is rather there to document one of the function's invariants, and possibly provided as an optimization hint to compilers in release builds. It's not an explicit check because this is not an assert that the user should be able to trip in any way.